pkg/controller/jobframework/reconciler.go

```diff
@@ -103,14 +104,31 @@ func NewReconciler(
 
 func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Request, job GenericJob) (ctrl.Result, error) {
 	object := job.Object()
-	if err := r.client.Get(ctx, req.NamespacedName, object); err != nil {
-		// we'll ignore not-found errors, since there is nothing to do.
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+	log := ctrl.LoggerFrom(ctx).WithValues("job", req.String())
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	err := r.client.Get(ctx, req.NamespacedName, object)
+
+	if apierrors.IsNotFound(err) || !object.GetDeletionTimestamp().IsZero() {
```

Comment:
Could you provide maybe first some context for what is the scenario we are fixing here? I was thinking the scenario is that a user deletes a workload, thus losing the node selectors. However, this if appears to be handling deletion of a Job, so I'm not sure how this is connected.

In particular, this code has user-facing impact. Please explain in the release note the impact, the fixed scenario.

Initially, I was thinking there would be a finalizer added to workloads, at workloads creation. Then, there would be a check in the workload_controller to see if the OwnerReference (the Job) exists. If the Job exists, then we can patch the workload to remove the finalizer.

Reply:
I see. Still, we could check if the owner Job exists or is being deleted (as here) in the workload_controller. Needing to deal with only a single owner job seems simpler.

To the second part, keeping the finalizer for the duration of the job existence seems simpler conceptually, and fewer API requests on suspend / unsuspend, so I'm trying to understand the motivation to delete the workload on suspend. Also, I'm not sure about keeping the workload until the  nodeSelectors are restored. How do we know if this happened? I guess we would need to have some condition to indicate that the nodeSelectors are restored. Otherwise relying just on suspend=true may give wrong results. WDYT?

Reply:
> I guess we would need to have some condition to indicate that the nodeSelectors are restored. Otherwise relying just on suspend=true may give wrong results. 

Given how Job requires 3 different API calls to suspend, it's indeed not enough to rely on the suspend field.
It's cleaner to just keep the finalizer until the Job is deleted (has a deletionTimestamp) or the Workload has a Finished condition.

Reply:
Initially , the presence of the finalizer for in the workload of a suspended job was an indicator that `stopJob` failed somewhere mid execution. However #1005 has another approach.

Reply:
What is the user-facing difference between:
1. workload lifetime is just for the running job
2. workload lifetime tied to the Job lifetime, for as long as they are equivalent

In (1.) we need two additional calls to add and remove the finalizer, so the performance is degraded (it can be argued how much). Are there drawbacks of (2.)?

Reply:
with 2, deleting a workload while the job is still present requires the job controller to remove the finalizer (one additional reconcile cycle, if triggered from outside the job controller)

Reply:
We could remove the finalizer from Workload in 2 scenarios:
- The Workload is already marked as finished.
- The Job is deleted.

Reply:
But still need to be removed if not finished and the job is suspended if the workload gets deleted.

Still the current implementation is the best fit, in my mind.

Reply:
> In particular, in the scenario of manual workload deletion (after suspending) it requires 2 API calls (one to add and one to delete the finalizer).

I don't follow.

Reply:
I had in mind that you require 2 API calls [here](https://github.com/kubernetes-sigs/kueue/blob/46980b39077eb773a08f48c4712c2fac622b0c41/pkg/controller/jobframework/reconciler.go#L474) and [here](https://github.com/kubernetes-sigs/kueue/blob/46980b39077eb773a08f48c4712c2fac622b0c41/pkg/controller/jobframework/reconciler.go#L522).
So, when you mentioned workload deletion I was thinking about the scenario:
1. Job created [here is the first call to add the finalizer]
2. job started
3. job suspended, workload unadmitted [+ API  call to remove the finalizer]
4. workload manually deleted by user, and replacement workload created [+ API call to add the finalizer, and +API call to eventually delete it]

Reply:
I guess, you had a different scenario, that the Job is never started, and the workload is manually deleted, right? I see, in this case the method you propose would not require any extra calls vs. one in the alternative. Still, the scenario seems much less important. Anyway, I will defer to @alculquicondor for the decision.

Reply:
To me, performance (in this case, less API calls in the happy path) and simplicity takes higher priority than covering an edge use case.

Reply:
The main purpose of this change is to make sure that when a job is unsuspended and its workload gets removed we still get the chance to restore it's node selectors. That is why I choose to "lock" the deletion for the time the job is unsuspended only. (In the previous iterations, before #1005, this approach was also useful in detecting mid stop errors ).

In an ideal life-cycle of a job/workload  (the workload is created once, is admitted and the job finishes) this will have one extra API call then the case when we set the finalizer at the time of creating the workload.  The number of API calls will change in favor of this approach if the workload gets recreated and against is if the workload gets preempted.

However changed so the finalizer is set from the workloads creation until it gets to `Finished`.

Reply:
> Not really. The finalizer can be included in the initial spec sent to create the Workload. This is not only one less API call, but overall less prone to race conditions.

That is the one I'm talking about, the removal needs to be done in a different call. (if it gets to finish, or if it gets deleted before that).

Reply:
Right. The point is that in this approach we no longer need an extra API call when admitting a workload.

Reply:
But need one if/when the workload gets deleted before being admitted (if the job spec changes). anyway....